### PR TITLE
Refactored listing resources to use React

### LINKED ISSUE #455: Rewrite listing page in React
This would have several benefits:
- No operations would require a page refresh anymore
- Easier to test listing page (currently there are no tests covering this part)
- Cleaner code
- Easier integration with existing React components

Potential complications:
- Integration with haystack text caching

Any thoughts?

### PR DESCRIPTION
**Pull request by noisecapella**
Fixes #455

### CODE REVIEW

File: ui/static/ui/js/listing.js [hunk 1]
```diff
@@ -322,27 +289,27 @@ define('listing',
...
-          showUpdateAllMembers();
-        })
-        .fail(function(data) {
-          //show alert
-          var message = 'Error adding user ' + email +
-            ' to group ' + formatGroupName(groupType);
-          message = message + '<br>' + data.responseJSON.username[0];
-          showMembersAlert(message, 'danger');
-        });
+          .done(function () {
```
Comment (giocalitri) [inline ui/static/ui/js/listing.js]: Why extra 2 white spaces? Is this the convention for chained function?

Comment (noisecapella) [inline ui/static/ui/js/listing.js]: This is just how the IDE reformatted it. I'll unindent this section

File: ui/views.py [hunk 2]
```diff
@@ -241,24 +243,50 @@ def extra_context(self):
...
+                "resource_type": result.resource_type,
+                "title": result.title,
+                "course": result.course,
+                "run": result.run,
+                "description": result.description,
+                "description_path": result.description_path,
+                "preview_url": result.preview_url,
+                "xa_nr_views": result.nr_views,
+                "xa_nr_attempts": result.nr_attempts,
+                "xa_avg_grade": result.avg_grade,
```
Comment (giocalitri) [inline ui/views.py:257]: why you are changing the name in the dict?

```
"xa_nr_views": result.nr_views,
"xa_nr_attempts": result.nr_attempts,
"xa_avg_grade": result.avg_grade,
```

Comment (noisecapella) [inline ui/views.py:257]: @ShawnMilo indicated that he was going to change it soon and I wanted to save the effort of doing it twice

Comment (giocalitri) [inline ui/views.py:257]: makes sense

**Comment by giocalitri** [comment]
Testing the functionality I found that clicking on the page number does not change it to a input box.

**Comment by giocalitri** [comment]
The number in the export button is not reset after exporting the learning resources

**Comment by giocalitri** [comment]
The rest of the functionality looks fine to me, but I would like a complete review from @pwilkins before merging.

**Comment by noisecapella** [comment]
I've addressed those comments. One other thing to look out for, I've changed the export button behavior to address #487, it will only show up when there are exports ready to export.

**Comment by giocalitri** [comment]
Weird behavior/bug (not sure is related to this PR, so if it isn't we need to open a bug):
if I select few learning resources and then in the panel I remove all of them, I can still download a file (empty, so it's not a big deal).

**Comment by giocalitri** [comment]
Travis is still broken, so you need to fix the problem. 
Other than that, if the above comment is unrelated to this PR, this is good to be merged.

File: ui/jstests/test_listing_resources.jsx [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        var resource = sampleResources[1];
+
+        var afterMount = function(component) {
+          // ListingResource has no state
+          assert.equal(null, component.state);
+
+          // Assert link behavior
+          var node = React.findDOMNode(component);
+          assert.deepEqual(openResourcePanelCount, {});
+          React.addons.TestUtils.Simulate.click($(node).find("h2 a")[0]);
```
Comment (amir-qayyum-khan) [inline ui/jstests/test_listing_resources.jsx]: Can you use a new variable to get  reference `$(node).find("h2 a")[0]`
This is suggestion to make code more clear.

File: ui/static/ui/js/listing.js [hunk 4]
```diff
@@ -1,162 +1,163 @@
...
-        }
-      });
-    }
+              pageNum,
+              maxPages,
+              $replaceWith,
+              $callerSelector
+            );
+        });
+        $($replaceWith).keypress(function(e) {
```
Comment (amir-qayyum-khan) [inline ui/static/ui/js/listing.js]: I think this is little confusing `$($xyz)`

**Comment by noisecapella** [comment]
Comments are addressed, I'll merge if tests pass

**Comment by giocalitri** [comment]
Please squash before merging

**Pull request merged**